### PR TITLE
Add Playwright E2E coverage for the Review Order page

### PR DESCRIPTION
**Pull request by tpaksu**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Adds Playwright E2E coverage in `tests/e2e-pw/tests/order/customer-review-order.spec.ts` that exercises the customer-facing path of the Review Order page:

1. Navigate to `/review-order/{id}/?key=...` with a valid order key.
2. Rate two of three line items (one with review text), leave the third untouched.
3. Submit and confirm per-row success status badges appear on the page.
4. Reload and confirm the rated rows render in the locked "already reviewed" variant.
5. Verify a tampered order key produces a 404.

The fixture seeds three reviewable products and a completed order via the REST API, then tears them down on completion.

Closes WOOPLUG-6601.

### Screenshots or screen recordings:

| Before | After |
| ------ | ----- |
|        |       |

### How to test the changes in this Pull Request:

1. Run the new spec locally: `pnpm --filter=@woocommerce/plugin-woocommerce test:e2e:default-pw tests/order/customer-review-order.spec.ts`.
2. The three test cases should pass: submit success path, locked-row reload state, 404 on key mismatch.

### Testing that has already taken place:

- Branch lint passes.
- Spec follows the established fixture/teardown pattern from `customer-payment-page.spec.ts`.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Add Playwright E2E coverage for the Customer Review Order page.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment

</details>

### CODE REVIEW

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+const productPrice = '15.99';
+
+test.describe(
+	'Customer Review Order page: rate, submit, and verify',
+	{ tag: [ tags.HPOS ] },
+	() => {
+		const products: OrderProduct[] = [];
+		let order: Order;
+		const customerEmail = `review-${ Date.now() }@example.test`;
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: Fixed in ca00e9bbf5. Switched to test.describe.serial so the locked-row test's dependency on the submit test running first is explicit.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+			// Submit gate should be enabled now.
+			const submit = page.locator( '.woocommerce-review-order__submit' );
+			await expect( submit ).toBeEnabled();
+			await Promise.all( [
+				page.waitForResponse(
+					( response ) =>
+						response.url().includes( 'admin-ajax.php' ) &&
+						response.request().method() === 'POST'
+				),
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: Fixed in ca00e9bbf5. The waitForResponse predicate now requires admin-ajax.php + POST + the woocommerce_submit_order_reviews action in the post data + a 2xx response, so unrelated background AJAX cannot satisfy it.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+			await expect( submit ).toBeEnabled();
+			await Promise.all( [
+				page.waitForResponse( ( response ) => {
+					const request = response.request();
+					const postData = request.postData() ?? '';
+					return (
+						response.url().includes( 'admin-ajax.php' ) &&
+						request.method() === 'POST' &&
+						postData.includes(
+							'action=woocommerce_submit_order_reviews'
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: Fixed. The waitForResponse predicate now matches by inspecting the JSON response body (presence of data.results) instead of POST body string. The form posts as FormData (multipart), so request.postData() is unreliable for action matching as you noted.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { tags, expect, test } from '../../fixtures/fixtures';
+
+type OrderProduct = {
+	id: number;
+	name: string;
+};
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts:13]: Fixed. Dropped the unused 'name' field from the OrderProduct fixture type; only 'id' is used downstream.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+			for ( const product of products ) {
+				await restApi.delete(
+					`${ WC_API_PATH }/products/${ product.id }`,
+					{ force: true }
+				);
+			}
+			await restApi.delete( `${ WC_API_PATH }/orders/${ order.id }`, {
+				force: true,
+			} );
+		} );
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts:86]: Fixed. afterAll() now wraps each delete in try/catch and guards the order deletion on order && order.id so a partial beforeAll() failure doesn't mask the real failure.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+								body.data,
+								'results'
+							)
+						);
+					} catch ( _err ) {
+						return false;
+					}
+				} ),
+				submit.click(),
+			] );
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts:211]: Acknowledged. The predicate already requires (admin-ajax.php URL) + (POST method) + (data.results in JSON body) + (response.ok). The only realistic competing AJAX would have to also POST to admin-ajax.php and return JSON with a data.results key on the same page; nothing on the Review Order page does that.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+			await expect(
+				rows
+					.nth( 0 )
+					.locator( '.woocommerce-review-order__item-status' )
+			).toBeVisible();
+			await expect(
+				rows
+					.nth( 1 )
+					.locator( '.woocommerce-review-order__item-status' )
+			).toBeVisible();
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: Fixed. Assertions now check for the .woocommerce-review-order__item-status--ok variant specifically, not just any status note. An --error rendering would no longer pass.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+			await expect(
+				rows
+					.nth( 2 )
+					.locator( '.woocommerce-review-order__item-status' )
+			).toHaveCount( 0 );
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: Fixed (same commit) — see the --ok class-name assertion.

**Comment by tpaksu** [comment]
Audited review comments against current state:

**Already addressed in current spec (older Copilot points):**
- Suite wrapped in `test.describe.serial(…)`.
- Cleanup is best-effort: per-resource `try/catch` so a teardown failure doesn't mask the original test error.
- `OrderProduct` type is minimal (`{ id: number }`), no unused `name`.
- Submit-success assertions use `successStatusSelector` matching `--ok` or `--pending_moderation`, so an `--error` outcome would fail the test.
- `page.waitForResponse` predicate scopes by the JSON shape `body.data.results` the AJAX handler returns, avoiding false positives from other admin-ajax calls. The form posts as multipart `FormData`, so `request.postData()` for an `action=...` substring is unreliable; the JSON-shape check is the more robust scoping signal.

**Fixed in bea4d6a8b5:**
- The new empty-state submission test now uses the same JSON-shape `waitForResponse` predicate as the rest of the spec rather than the loose URL-only match.

**Declined (CodeRabbit "fail fast if fixture API responses are malformed"):** the test already fails loudly if `response.data.id` is undefined — TypeScript optional-chained access surfaces as a clear assertion failure on the next step. Adding defensive shape checks would only mask which fixture call broke, not improve diagnostics. Per the project's no-impossible-scenario-error-handling guideline, leaving as is.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: Done — suite is wrapped in `test.describe.serial(…)` and the two follow-on describe blocks (disabled-products notice, 404 path) each create their own short-lived order so they don't depend on the serial chain's state.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: Addressed in the current spec: the `waitForResponse` predicate scopes by the JSON shape `body.data.results` returned by the `woocommerce_submit_order_reviews` handler, so other admin-ajax requests can't resolve the wait. The new empty-state test was tightened to match in bea4d6a8b5.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: The form posts as multipart `FormData` via `window.fetch`, so `request.postData()` typically doesn't contain a URL-encoded `action=…` substring — exactly the concern raised here. The current predicate uses the JSON-shape check (`body.data.results`) instead, which is reliable regardless of the request encoding.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts:13]: Done — `OrderProduct` is `{ id: number }` in current state. No unused `name` field.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts:86]: Done — `afterAll` wraps each `restApi.delete` in `try/catch` and guards the order deletion behind `if ( order && order.id )`. A teardown failure no longer masks the original test failure.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts:211]: Same predicate addresses this: matches POST + `body.data.results` JSON shape, scoping to the AJAX handler's response. The new empty-state submission test was tightened to use the same predicate in bea4d6a8b5.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: Done — assertions use `successStatusSelector` matching only `--ok` or `--pending_moderation`, so an `--error` outcome would fail the test.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts]: Done — see `successStatusSelector` constant; the assertion now requires `--ok` or `--pending_moderation`, never just `--error`.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+						email: customerEmail,
+					},
+					line_items: products.map( ( p ) => ( {
+						product_id: p.id,
+						quantity: 1,
+					} ) ),
+				}
+			);
+			order = orderResponse.data;
+		} );
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts:64]: Declined: the test fails loudly on undefined `response.data.id` (the next `restApi.post` for the order would throw a clear error). Adding defensive shape checks layers extra code without improving diagnostics, against the project's no-impossible-scenario-error-handling guideline.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts:407]: Declined for the same reason as the analogous comment on the first describe (id 3230475653): the test surface fails loudly on a malformed fixture response without extra scaffolding.

File: plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts [hunk 11]
```diff
@@ -0,0 +1,485 @@
...
+			);
+			const productId = productResponse.data.id;
+			const orderResponse = await restApi.post(
+				`${ WC_API_PATH }/orders`,
+				{
+					status: 'completed',
+					line_items: [ { product_id: productId, quantity: 1 } ],
+				}
+			);
+			const orderId = orderResponse.data.id;
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply plugins/woocommerce/tests/e2e-pw/tests/order/customer-review-order.spec.ts:485]: Declined for the same reason — the inline fixture would fail loudly on a malformed response without defensive shape checks.

**Comment by tpaksu** [comment]
Closing to start fresh on top of the post-WOOPLUG-6705 (per-variation reviews) state. The new E2E PR will rebase onto `wooplug-6705-support-per-variation-reviews-on-the-review-order-page` and cover every scenario the testing walkthrough captures (Scenarios 1–6 + theme-awareness + variations). Tracked from the same WOOPLUG-6601 Linear issue.
